File: problem_009.c
// -*- compile-command: "gcc -std=c99 -o problem_009 problem_009.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    const int target = 1000;

    printf("Project Euler - Problem 9:\n"
           "There exists exactly one Pythagorean triplet for which a + b + c = 1000. Find the product abc.\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int product = 0;
    int a = 1;
    while (a++)
    {
        int b = a;
        while (b++)
        {
            int c = sqrt(a*a + b*b);
            int sum = a + b + c;
            if (sum > target) break;
            if (c != sqrt(a*a + b*b)) continue;
            if (sum == target)
            {
                product = a * b * c;
                break;
            }
        }
        if (product != 0) break;
    }

    printf("Product: %d\n", product);
    printElapsedTime(start);

    return 0;
}

File: problem_009.rs
// -*- compile-command: "rustc -o problem_009_rs problem_009.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 9");
    println!("There exists exactly one Pythagorean triplet for which a + b + c = 1000. Find the product abc.\n");

    let mut answer = 0;

    for a in 1..1000 {
        for b in a+1..1000 {
            let c = ((a*a + b*b) as f64).sqrt();
            if c <= b as f64 {
                continue;
            }
            if a as f64+b as f64+c == 1000 as f64 {
                //println!("{} {} {}", a, b, c);
                answer = a*b*c as u32;
                break;
            }
        }
        if answer != 0 {
            break;
        }
    }

    println!("{}", answer);
}
